### ISSUE TITLE
Haproxy stuck at 100% cpu usage before restart

### ISSUE THREAD

**Issue by ut0mt8** [haproxy/haproxy#2169]
### Detailed Description of the Problem

A little context before. 
We use haproxy in our company as a "poor-man" replacement of managed LB on AWS. 
The reason we did that is two fold: 
- for saving cost reason mostly as managed LB cost will kill us
- also for intercepting response code as we basically want to answer 204 on every error
we are running a SSP platform (following openRTB standard) with some trafic (~2 millions RPS in peak) distributed over multiple haproxy with some kind of DNS load-balancing magic (route53 stuffz).
this is far from perfect but it do the job.

my problem is that sometimes on one instance Haproxy seems to enter a deadloop eating 100% cpu of the box.
what is strange is that just restarting haproxy solve the issue.

 

### Expected Behavior

Haproxy not entering this state.

### Steps to Reproduce the Behavior

not easy. I supposed we have to reproduce the behaviour of my clients ?

### Do you have any idea what may have caused this?

when it occurs it seems that both number of sessions rates / and number of sessions is too high for this poor instance.

### Do you have an idea how to solve the issue?

not at the moment; maybe lower the limits/ maybe adding random restart / close on my instance to spread connections?

### What is your configuration?

```` 
global
  maxconn 100000
  log 127.0.0.1 local0 notice
  stats timeout 10s
  stats socket /run/haproxy-admin.sock mode 660 level admin expose-fd listeners
  tune.ssl.default-dh-param 1024
  tune.ssl.cachesize 2000000
  tune.bufsize 262144

defaults
  log global
  option dontlognull
  option dontlog-normal
  timeout connect 1s
  timeout client 10s
  timeout server 10s
  timeout client-fin 1s
  timeout server-fin 1s
  timeout http-keep-alive 180s
  retries 0
  option splice-auto
  option http-keep-alive
  http-reuse aggressive
  option http-buffer-request

listen stats
  bind *:9101
  mode http
  http-request use-service prometheus-exporter if { path /metrics }
  stats enable
  stats uri /
  stats show-node
  stats show-modules
  stats refresh 25s

http-errors custom_errors
  errorfile 500 /etc/haproxy/204.http
  errorfile 502 /etc/haproxy/204.http
  errorfile 503 /etc/haproxy/204.http
  errorfile 504 /etc/haproxy/204.http

frontend http_https_in
  maxconn 100000
  bind *:80 name http
  bind *:443 ssl crt /etc/haproxy/star.dsp.corp.com.pem name https
  mode http
  compression algo gzip
  compression type text/css text/html text/javascript application/javascript text/plain text/xml application/json
  monitor-uri /health
  monitor fail if { nbsrv(bidder) eq 0 }
  default_backend bidder

backend prefiltering
  mode http
  http-request return status 204

backend bidder
  mode http
  errorfiles custom_errors
  http-response return status 500 default-errorfiles if { status ge 400 }
  http-request add-header X-Forwarded-Proto https if { ssl_fc }
  balance roundrobin
  default-server inter 5s fall 3 rise 2 downinter 240s slowstart 45s check port 1331 resolve-prefer ipv4 agent-check agent-port 1935 agent-inter 2s weight 100

  server 10.x.x.x 10.x.x.x:5000
  ...
````

### Output of `haproxy -vv`

````
Status: stable branch - will stop receiving fixes around Q1 2024.
Known bugs: http://www.haproxy.org/bugs/bugs-2.7.4.html
Running on: Linux 6.1.19-30.43.amzn2023.aarch64 #1 SMP Wed Mar 15 14:44:03 UTC 2023 aarch64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY +LUA -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT +PCRE -PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC +RT +SHM_OPEN -SLZ -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL +ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=8).
Built with OpenSSL version : OpenSSL 3.0.8 7 Feb 2023
Running on OpenSSL version : OpenSSL 3.0.8 7 Feb 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.4
Built with the Prometheus exporter as a service
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.44 2020-02-12
Running on PCRE version : 8.44 2020-02-12
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.3.1 20221121 (Red Hat 11.3.1-4)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace

````

### Last Outputs and Backtraces

````
echo "show info" | socat stdio unix-connect:/run/haproxy-admin.sock
Name: HAProxy
Version: 2.7.4-d28541d
Release_date: 2023/03/10
Nbthread: 8
Nbproc: 1
Process_num: 1
Pid: 2057
Uptime: 1d 1h50m17s
Uptime_sec: 93017
Memmax_MB: 0
PoolAlloc_MB: 61
PoolUsed_MB: 61
PoolFailed: 0
Ulimit-n: 250575
Maxsock: 250575
Maxconn: 100000
Hard_maxconn: 100000
CurrConns: 37572
CumConns: 109267798
CumReq: 1814640550
MaxSslConns: 0
CurrSslConns: 26587
CumSslConns: 73571396
Maxpipes: 25000
PipesUsed: 0
PipesFree: 0
ConnRate: 2783
ConnRateLimit: 0
MaxConnRate: 53572
SessRate: 2776
SessRateLimit: 0
MaxSessRate: 52620
SslRate: 2660
SslRateLimit: 0
MaxSslRate: 7663
SslFrontendKeyRate: 0
SslFrontendMaxKeyRate: 1239
SslFrontendSessionReuse_pct: 100
SslBackendKeyRate: 0
SslBackendMaxKeyRate: 0
SslCacheLookups: 11
SslCacheMisses: 11
CompressBpsIn: 20918
CompressBpsOut: 6968
CompressBpsRateLim: 0
ZlibMemUsage: 0
MaxZlibMemUsage: 0
Tasks: 38698
Run_queue: 26713
Idle_pct: 0
node: xxx
Stopping: 0
Jobs: 37692
Unstoppable Jobs: 1
Listeners: 5
ActivePeers: 0
ConnectedPeers: 0
DroppedLogs: 0
BusyPolling: 0
FailedResolutions: 0
TotalBytesOut: 5183281048152
TotalSplicdedBytesOut: 0
BytesOutRate: 29317536
DebugCommandsIssued: 0
CumRecvLogs: 0
Build info: 2.7.4-d28541d
Memmax_bytes: 0
PoolAlloc_bytes: 64481368
PoolUsed_bytes: 64481368
Start_time_sec: 1684677333
Tainted: 0
````
````

echo "show pools" | socat stdio unix-connect:/run/haproxy-admin.sock
Dumping pools usage. Use SIGQUIT to flush them.
  - Pool cache_st (24 bytes) : 161 allocated (3864 bytes), 161 used (~0 by thread caches), needed_avg 158, 0 failures, 5 users, @0x326e66c0 [SHARED]
  - Pool hlua_flt_ct (56 bytes) : 42887 allocated (2401672 bytes), 42887 used (~5 by thread caches), needed_avg 42883, 0 failures, 13 users, @0x326c1f40 [SHARED]
  - Pool stconn (88 bytes) : 625 allocated (55000 bytes), 625 used (~0 by thread caches), needed_avg 621, 0 failures, 8 users, @0x326ea780 [SHARED]
  - Pool ssl_sock_ct (120 bytes) : 29916 allocated (3589920 bytes), 29916 used (~7 by thread caches), needed_avg 29917, 0 failures, 1 users, @0x326b9dc0 [SHARED]
  - Pool hlua (152 bytes) : 95195 allocated (14469640 bytes), 95195 used (~14 by thread caches), needed_avg 95192, 0 failures, 4 users, @0x326bde80 [SHARED]
  - Pool h2s (184 bytes) : 53242 allocated (9796528 bytes), 53242 used (~8 by thread caches), needed_avg 53238, 0 failures, 6 users, @0x326c6040 [SHARED]
  - Pool tcpcheck_ru (216 bytes) : 41469 allocated (8957304 bytes), 41469 used (~5 by thread caches), needed_avg 41466, 0 failures, 5 users, @0x326d2280 [SHARED]
  - Pool ssl_sock_cl (280 bytes) : 1 allocated (280 bytes), 1 used (~0 by thread caches), needed_avg 0, 0 failures, 3 users, @0x326b5d00 [SHARED]
  - Pool spoe_ctx (312 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x326da440 [SHARED]
  - Pool resolv_reso (472 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x326de500 [SHARED]
  - Pool resolv_answ (600 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x326e25c0 [SHARED]
  - Pool stream (984 bytes) : 162 allocated (159408 bytes), 162 used (~0 by thread caches), needed_avg 159, 0 failures, 1 users, @0x326d6380 [SHARED]
  - Pool requri (1048 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x328f1d40 [SHARED]
  - Pool fcgi_conn (1240 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x326ce1c0 [SHARED]
  - Pool h2c (1304 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x326ca100 [SHARED]
  - Pool hpack_tbl (4096 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x327ab5c0 [SHARED]
  - Pool zlib_state (5976 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x327a7500 [SHARED]
  - Pool dns_msg_buf (65560 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 1, 0 failures, 5 users, @0x326ee840 [SHARED]
  - Pool buffer (262144 bytes) : 134 allocated (35127296 bytes), 134 used (~7 by thread caches), needed_avg 130, 0 failures, 1 users, @0x32933c80 [SHARED]
  - Pool trash (262176 bytes) : 1 allocated (262176 bytes), 1 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x326f2900
Total: 20 pools, 74823088 bytes allocated, 74823088 used (~1840808 by thread caches).
````

**Comment by Darlelet** [comment]
Thanks for the report

It would be nice to see if you still encounter the issue in 2.7.8: ~150 bugs were addressed since 2.7.4 (http://www.haproxy.org/bugs/bugs-2.7.4.html), including bugs that refer to 100% cpu usage due to spinning loops (e.g.: #2046)

**Comment by ut0mt8** [comment]
Sure let's give it a try.

**Comment by ut0mt8** [comment]
Well for now it's conclusive. But let's wait for a day or two before celebrate (and close the issue)

**Comment by Darlelet** [comment]
Great, thank you for the feedback.

> But let's wait for a day or two before celebrate (and close the issue)

Yeah sure, no problem!

**Comment by ut0mt8** [comment]
Unfortunately the behaviour reappear (seems a bit sparse) but still. 

the configuration slightly change : 

`````
  timeout server 1s (10 before)
`````

`````
HAProxy version 2.7.8-58c657f 2023/05/02 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2024.
Known bugs: http://www.haproxy.org/bugs/bugs-2.7.8.html
Running on: Linux 6.1.21-1.45.amzn2023.aarch64 #1 SMP Wed Mar 29 00:52:08 UTC 2023 aarch64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY +LUA -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT +PCRE -PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC +RT +SHM_OPEN -SLZ -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL +ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=8).
Built with OpenSSL version : OpenSSL 3.0.8 7 Feb 2023
Running on OpenSSL version : OpenSSL 3.0.8 7 Feb 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.4
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.44 2020-02-12
Running on PCRE version : 8.44 2020-02-12
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.3.1 20221121 (Red Hat 11.3.1-4)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
`````

`````
[root@ip-10-80-250-241 ec2-user]# echo "show info" | socat stdio unix-connect:/run/haproxy-admin.sock
Name: HAProxy
Version: 2.7.8-58c657f
Release_date: 2023/05/02
Nbthread: 8
Nbproc: 1
Process_num: 1
Pid: 2044
Uptime: 1d 1h55m23s
Uptime_sec: 93323
Memmax_MB: 0
PoolAlloc_MB: 92
PoolUsed_MB: 92
PoolFailed: 0
Ulimit-n: 250575
Maxsock: 250575
Maxconn: 100000
Hard_maxconn: 100000
CurrConns: 47395
CumConns: 101762777
CumReq: 2180279299
MaxSslConns: 0
CurrSslConns: 31196
CumSslConns: 63475336
Maxpipes: 25000
PipesUsed: 0
PipesFree: 0
ConnRate: 3687
ConnRateLimit: 0
MaxConnRate: 56179
SessRate: 3671
SessRateLimit: 0
MaxSessRate: 56049
SslRate: 3464
SslRateLimit: 0
MaxSslRate: 8323
SslFrontendKeyRate: 0
SslFrontendMaxKeyRate: 2725
SslFrontendSessionReuse_pct: 100
SslBackendKeyRate: 0
SslBackendMaxKeyRate: 0
SslCacheLookups: 13
SslCacheMisses: 13
CompressBpsIn: 69856
CompressBpsOut: 23033
CompressBpsRateLim: 0
ZlibMemUsage: 0
MaxZlibMemUsage: 0
Tasks: 48975
Run_queue: 31446
Idle_pct: 0
node: ip-10-80-250-241.ap-southeast-1.compute.internal
Stopping: 0
Jobs: 47510
Unstoppable Jobs: 1
Listeners: 5
ActivePeers: 0
ConnectedPeers: 0
DroppedLogs: 0
BusyPolling: 0
FailedResolutions: 0
TotalBytesOut: 6397419303373
TotalSplicdedBytesOut: 0
BytesOutRate: 51254528
DebugCommandsIssued: 0
CumRecvLogs: 0
Build info: 2.7.8-58c657f
Memmax_bytes: 0
PoolAlloc_bytes: 97389768
PoolUsed_bytes: 97389768
Start_time_sec: 1685280683
Tainted: 0
`````

`````
echo "show pools" | socat stdio unix-connect:/run/haproxy-admin.sock
[root@ip-10-80-250-241 ec2-user]# echo "show pools" | socat stdio unix-connect:/run/haproxy-admin.sock
Dumping pools usage. Use SIGQUIT to flush them.
  - Pool cache_st (24 bytes) : 249 allocated (5976 bytes), 249 used (~4 by thread caches), needed_avg 252, 0 failures, 5 users, @0x28ebb640 [SHARED]
  - Pool hlua_flt_ct (56 bytes) : 56951 allocated (3189256 bytes), 56951 used (~36 by thread caches), needed_avg 56947, 0 failures, 13 users, @0x28e96f00 [SHARED]
  - Pool stconn (88 bytes) : 885 allocated (77880 bytes), 885 used (~12 by thread caches), needed_avg 883, 0 failures, 8 users, @0x28ebf700 [SHARED]
  - Pool ssl_sock_ct (120 bytes) : 37985 allocated (4558200 bytes), 37985 used (~15 by thread caches), needed_avg 37983, 0 failures, 1 users, @0x28e8ed40 [SHARED]
  - Pool hlua (152 bytes) : 128446 allocated (19523792 bytes), 128446 used (~58 by thread caches), needed_avg 128442, 0 failures, 4 users, @0x28e92e40 [SHARED]
  - Pool h2s (184 bytes) : 72756 allocated (13387104 bytes), 72756 used (~26 by thread caches), needed_avg 72752, 0 failures, 6 users, @0x28e9afc0 [SHARED]
  - Pool tcpcheck_ru (216 bytes) : 55246 allocated (11933136 bytes), 55246 used (~19 by thread caches), needed_avg 55245, 0 failures, 5 users, @0x28ea7240 [SHARED]
  - Pool ssl_sock_cl (280 bytes) : 1 allocated (280 bytes), 1 used (~0 by thread caches), needed_avg 0, 0 failures, 3 users, @0x28e8ac80 [SHARED]
  - Pool spoe_ctx (312 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x28eaf3c0 [SHARED]
  - Pool resolv_reso (472 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x28eb34c0 [SHARED]
  - Pool resolv_answ (600 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x28eb7580 [SHARED]
  - Pool stream (984 bytes) : 250 allocated (246000 bytes), 250 used (~4 by thread caches), needed_avg 253, 0 failures, 1 users, @0x28eab300 [SHARED]
  - Pool requri (1048 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x290c7d40 [SHARED]
  - Pool fcgi_conn (1240 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x28ea3180 [SHARED]
  - Pool h2c (1304 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x28e9f080 [SHARED]
  - Pool hpack_tbl (4096 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x28f805c0 [SHARED]
  - Pool zlib_state (5976 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x28f7c500 [SHARED]
  - Pool dns_msg_buf (65560 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 1, 0 failures, 5 users, @0x28ec3800 [SHARED]
  - Pool buffer (262144 bytes) : 163 allocated (42729472 bytes), 163 used (~3 by thread caches), needed_avg 161, 0 failures, 1 users, @0x29109c80 [SHARED]
  - Pool trash (262176 bytes) : 1 allocated (262176 bytes), 1 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x28ec78c0
Total: 20 pools, 95913272 bytes allocated, 95913272 used (~813040 by thread caches).

`````

**Comment by ut0mt8** [comment]
maybe this particular haproxy from my fleet is bombed by request (after all dns load balancing is completely unperfect) 
in this case how can I protect my instance?

**Comment by wtarreau** [comment]
I'm not noticing anything obviously wrong in your config. If just restarting haproxy addresses the problem, it would make sense to consider that haproxy is at fault here. Maybe there's a problem with some connections and closing them addresses it, I don't know. When this happens, the following would be useful:
```
- show activity
- show profiling tasks
```
Also, please run "perf top" on the machine to see where the CPU is spent. We might notice that it's mostly on a lock, or in the kernel, or in a function being called in loops, etc.

**Comment by ut0mt8** [comment]
Yup I suspect that as well ; some connections should make problem and restarting haproxy kill them, and possibly are being redispatched elsewhere. Here the trace asked.

````
echo "show activity" | socat stdio unix-connect:/run/haproxy-admin.sock
thread_id: 8 (1..8)
date_now: 1685451128.556188
uptime_now: 82777.821325
ctxsw: 856747387 [ 1723839672 1715192445 1716075796 1718867432 1715476098 1721860290 1716802377 1713535165 ]
tasksw: 4235807705 [ 531333301 528750586 529039988 529922013 528677237 530713612 529101345 528269623 ]
empty_rq: 186686735 [ 25946086 29367382 15052679 18990241 22510325 16648398 28184895 29986729 ]
long_rq: 19207221 [ 2411266 2402213 2399659 2400381 2401301 2395787 2394874 2401740 ]
loops: 3838372446 [ 484190387 485013892 470947405 475661540 478508251 474686989 484137225 485226757 ]
wake_tasks: 19228251 [ 2413832 2404948 2402053 2402870 2403761 2398299 2397560 2404928 ]
wake_signal: 0 [ 0 0 0 0 0 0 0 0 ]
poll_io: 3835861342 [ 483879652 484702361 470629748 475344284 478192417 474369222 483824583 484919075 ]
poll_exp: 423643 [ 53032 52803 53003 53282 53021 52860 52861 52781 ]
poll_drop_fd: 219901 [ 25117 28848 27645 26893 27192 28191 28181 27834 ]
poll_skip_fd: 0 [ 0 0 0 0 0 0 0 0 ]
conn_dead: 0 [ 0 0 0 0 0 0 0 0 ]
stream_calls: 4151263936 [ 520809903 518140564 518471121 519333572 518116084 520209164 518543413 517640115 ]
pool_fail: 0 [ 0 0 0 0 0 0 0 0 ]
buf_wait: 0 [ 0 0 0 0 0 0 0 0 ]
cpust_ms_tot: 809530 [ 27525 110268 113233 112447 112077 111164 112793 110023 ]
cpust_ms_1s: 16 [ 0 3 3 1 5 1 2 1 ]
cpust_ms_15s: 1093 [ 10 111 226 120 179 102 185 160 ]
avg_cpu_pct: 100 [ 100 100 100 100 100 100 100 100 ]
avg_loop_us: 7079 [ 7009 7127 6909 7055 7139 7216 7243 6936 ]
accepted: 76946626 [ 7328768 13147643 5999161 6178099 6523042 6030807 9014287 22724819 ]
accq_pushed: 76946626 [ 9610651 9634511 9609339 9612677 9616285 9618874 9619640 9624649 ]
accq_full: 0 [ 0 0 0 0 0 0 0 0 ]
accq_ring: 11 [ 0 9 0 0 1 1 0 0 ]
fd_takeover: 150325810 [ 18813023 18786086 18783656 18791337 18771617 18778782 18811253 18790056 ]
````

show profiling task show nothing.

perf top was indeed very interesting.

````
2.24%  libcrypto.so.3.0.8                            [.] 0x0000000000299a10
   1.72%  libcrypto.so.3.0.8                            [.] 0x00000000002999ec
   1.53%  libcrypto.so.3.0.8                            [.] 0x00000000000e6f2c
   1.26%  libcrypto.so.3.0.8                            [.] 0x0000000000299924
   1.19%  libcrypto.so.3.0.8                            [.] 0x000000000029a064
   1.14%  libcrypto.so.3.0.8                            [.] 0x000000000029a0ac
   1.12%  libcrypto.so.3.0.8                            [.] 0x000000000029967c
   1.05%  libcrypto.so.3.0.8                            [.] 0x00000000002998a4
   0.99%  libc.so.6                                     [.] _int_free
   0.87%  libcrypto.so.3.0.8                            [.] 0x0000000000299658
   0.87%  libcrypto.so.3.0.8                            [.] 0x00000000002999a8
   0.84%  libcrypto.so.3.0.8                            [.] 0x0000000000299a08
   0.78%  libcrypto.so.3.0.8                            [.] 0x0000000000299918
   0.76%  libcrypto.so.3.0.8                            [.] 0x00000000002999d0
   0.71%  libcrypto.so.3.0.8                            [.] 0x0000000000299914
   0.69%  libcrypto.so.3.0.8                            [.] 0x0000000000299910
   0.68%  libcrypto.so.3.0.8                            [.] 0x00000000002999e0
   0.67%  libcrypto.so.3.0.8                            [.] 0x00000000002999e8
   0.67%  libcrypto.so.3.0.8                            [.] 0x00000000002999d8
   0.67%  libcrypto.so.3.0.8                            [.] 0x00000000000e6f38
   0.67%  libc.so.6                                     [.] _int_malloc
   0.66%  libcrypto.so.3.0.8                            [.] 0x00000000002998ec
   0.65%  libcrypto.so.3.0.8                            [.] 0x00000000002998f4
   0.65%  libcrypto.so.3.0.8                            [.] 0x0000000000299920
   0.63%  libcrypto.so.3.0.8                            [.] 0x00000000002998fc
   0.62%  libcrypto.so.3.0.8                            [.] 0x00000000000e6f48
   0.62%  libcrypto.so.3.0.8                            [.] 0x00000000002998e4
   0.60%  libcrypto.so.3.0.8                            [.] 0x0000000000299810
   0.60%  libcrypto.so.3.0.8                            [.] 0x00000000002999c8
   0.59%  libcrypto.so.3.0.8                            [.] 0x0000000000299a04
   0.53%  libcrypto.so.3.0.8                            [.] 0x00000000002998d4
   0.53%  libcrypto.so.3.0.8                            [.] 0x0000000000299a0c
   0.53%  libcrypto.so.3.0.8                            [.] 0x00000000002999d4
   0.52%  libcrypto.so.3.0.8                            [.] 0x00000000002998f0
   0.52%  libcrypto.so.3.0.8                            [.] 0x00000000002999e4
   0.52%  libcrypto.so.3.0.8                            [.] 0x00000000002999dc
   0.52%  libcrypto.so.3.0.8                            [.] 0x00000000002998b4
   0.52%  libcrypto.so.3.0.8                            [.] 0x00000000002999cc
   0.51%  libcrypto.so.3.0.8                            [.] 0x000000000029991c
   0.51%  libcrypto.so.3.0.8                            [.] 0x0000000000299928
   0.51%  libcrypto.so.3.0.8                            [.] 0x00000000002998f8
   0.50%  libcrypto.so.3.0.8                            [.] 0x0000000000299900
   0.48%  libcrypto.so.3.0.8                            [.] 0x00000000002998e8
   0.47%  libcrypto.so.3.0.8                            [.] 0x00000000002998e0
   0.46%  libcrypto.so.3.0.8                            [.] 0x00000000002999b4
   0.46%  libcrypto.so.3.0.8                            [.] 0x00000000002998d8
   0.46%  libcrypto.so.3.0.8                            [.] 0x000000000029998c
   0.45%  libcrypto.so.3.0.8                            [.] 0x00000000002999ac
   0.43%  libcrypto.so.3.0.8                            [.] 0x0000000000299904
   0.42%  libcrypto.so.3.0.8                            [.] 0x00000000002999c0
   0.41%  libcrypto.so.3.0.8                            [.] 0x00000000002998dc
   0.40%  libcrypto.so.3.0.8                            [.] 0x0000000000299614
   0.40%  libcrypto.so.3.0.8                            [.] 0x0000000000299674
   0.39%  libcrypto.so.3.0.8                            [.] 0x00000000002998cc
   0.39%  libc.so.6                                     [.] malloc
   0.39%  libcrypto.so.3.0.8                            [.] 0x0000000000299e60
   0.38%  libcrypto.so.3.0.8                            [.] 0x00000000002999c4
   0.38%  libcrypto.so.3.0.8                            [.] 0x0000000000299a00
   0.38%  libcrypto.so.3.0.8                            [.] 0x000000000029963c
   0.38%  libcrypto.so.3.0.8                            [.] 0x00000000002999a4
   0.36%  libcrypto.so.3.0.8                            [.] 0x00000000000e6f58
   0.36%  libcrypto.so.3.0.8                            [.] 0x0000000000299f88
   0.35%  libcrypto.so.3.0.8                            [.] 0x000000000029a098
````

compared to other similar server where the result is well balanced between haproxy different kernel task.
so definitely it narrow down the scope of research.

**Comment by Darlelet** [comment]
Were the perf top and the various commands (show profiling tasks, activity..) taken during the surge (when haproxy is stuck at 100%)?

Also it seems that you used `show profiling tasks` without first enabling profiling, which could explain why nothing gets reported.

Here is how to use it (when haproxy is stuck at 100% cpu):

`set profiling tasks on`

Wait for a few seconds

Then

`set profiling tasks off`

Now you can get the report for the given observation period using `show profiling tasks`

It could be useful to compare tasks profiling reports for 2 instances enduring roughly the same traffic during (approximately) the same observation period (one healthy instance and the other with 100% cpu). This could point us in the right direction.

Also "show tasks" cli on the sick instance might be useful as well (this one is standalone, no need to activate / deactivate profiling)

**Comment by ut0mt8** [comment]
Yes the 

> Were the perf top and the various commands (show profiling tasks, activity..) taken during the surge (when haproxy is stuck at 100%)?
> 
> Also it seems that you used `show profiling tasks` without first enabling profiling, which could explain why nothing gets reported.
> 
> Here is how to use it (when haproxy is stuck at 100% cpu):
> 

Yes.

> `set profiling tasks on`
> 
understood

**Comment by ut0mt8** [comment]
ok here the "show profiling tasks" output : 

1 run on server stuck at 100%
`````
[root@ip-10-80-250-220 ec2-user]# echo "show profiling tasks" | socat stdio unix-connect:/run/haproxy-admin.sock
Tasks activity:
  function                      calls   cpu_tot   cpu_avg   lat_tot   lat_avg
  h1_io_cb                     835971   7.691s    9.200us   1.570m    112.7us <- sock_conn_iocb@src/sock.c:869 tasklet_wakeup
  process_stream               421054   6.573s    15.61us   48.44m    6.903ms <- sc_notify@src/stconn.c:1234 task_wakeup
  sc_conn_io_cb                419592   769.9ms   1.834us   1.619m    231.5us <- h1_wake_stream_for_recv@src/mux_h1.c:2573 tasklet_wakeup
  sc_conn_io_cb                418259   158.9ms   379.0ns   1.806m    259.1us <- sc_app_chk_rcv_conn@src/stconn.c:789 tasklet_wakeup
  process_stream               418126   7.592s    18.16us   47.53m    6.820ms <- stream_new@src/stream.c:573 task_wakeup
  h1_io_cb                      79897   25.95ms   324.0ns   4.621m    3.470ms <- h1_takeover@src/mux_h1.c:4124 tasklet_wakeup
  accept_queue_process          41049   863.6ms   21.04us   1.989m    2.907ms <- listener_accept@src/listener.c:1209 tasklet_wakeup
  ssl_sock_io_cb                38469   2.660m    4.149ms   14.92h    1.396s  <- ssl_sock_start@src/ssl_sock.c:5864 tasklet_wakeup
  session_expire_embryonic       8815   190.5ms   21.61us   3.594s    407.7us <- wake_expired_tasks@src/task.c:350 task_wakeup
  h1_io_cb                       5100   113.7ms   22.30us   304.7ms   59.74us <- conn_subscribe@src/connection.c:736 tasklet_wakeup
  process_chk                    4125   72.96ms   17.69us   1.953s    473.5us <- wake_expired_tasks@src/task.c:350 task_wakeup
  h1_timeout_task                3521   59.71ms   16.96us   1.178s    334.5us <- wake_expired_tasks@src/task.c:350 task_wakeup
  accept_queue_process           1676   5.776ms   3.446us   1.046s    624.1us <- accept_queue_process@src/listener.c:166 tasklet_wakeup
  h1_timeout_task                1591   1.080ms   678.0ns   9.643s    6.061ms <- h1_release@src/mux_h1.c:1041 task_wakeup
  process_chk                     205   3.462ms   16.89us   1.052s    5.131ms <- wake_srv_chk@src/check.c:1055 task_wakeup
  sc_conn_io_cb                   193   1.954ms   10.12us   72.47ms   375.5us <- h1_wake_stream_for_send@src/mux_h1.c:2583 tasklet_wakeup
  h1_io_cb                        193   422.2us   2.187us   54.93ms   284.6us <- sock_conn_iocb@src/sock.c:849 tasklet_wakeup
  srv_chk_io_cb                   145   829.5us   5.720us   24.38ms   168.2us <- sock_conn_iocb@src/sock.c:849 tasklet_wakeup
  srv_chk_io_cb                   144   1.162ms   8.071us   6.045ms   41.98us <- sock_conn_iocb@src/sock.c:869 tasklet_wakeup
  other                            76      -         -      755.7ms   9.943ms <- wake_expired_tasks@src/task.c:350 task_wakeup
  srv_cleanup_toremove_conns       36   617.6us   17.16us   145.4ms   4.039ms <- srv_cleanup_idle_conns@src/server.c:5976 task_wakeup
  srv_cleanup_idle_conns           20   57.52us   2.876us   6.051ms   302.6us <- wake_expired_tasks@src/task.c:435 task_drop_running
  task_run_applet                   6   9.608ms   1.601ms   5.293ms   882.2us <- run_tasks_from_lists@src/task.c:658 task_drop_running
  task_run_applet                   3   2.124ms   708.0us   6.109ms   2.036ms <- sc_applet_create@src/stconn.c:496 appctx_wakeup
  task_run_applet                   1   1.864ms   1.864ms   9.157us   9.157us <- sc_app_chk_rcv_applet@src/stconn.c:994 appctx_wakeup
`````

2 run on server acting normaly
`````
Tasks activity:
  function                      calls   cpu_tot   cpu_avg   lat_tot   lat_avg
  h1_io_cb                     680985   4.218s    6.193us   743.4ms   1.091us <- sock_conn_iocb@src/sock.c:869 tasklet_wakeup
  process_stream               444730   5.811s    13.07us   1.224s    2.752us <- sc_notify@src/stconn.c:1234 task_wakeup
  sc_conn_io_cb                444571   776.1ms   1.745us   368.8ms   829.0ns <- h1_wake_stream_for_recv@src/mux_h1.c:2573 tasklet_wakeup
  sc_conn_io_cb                325064   169.6ms   521.0ns   3.562s    10.96us <- sc_app_chk_rcv_conn@src/stconn.c:789 tasklet_wakeup
  process_stream               309498   3.110s    10.05us   1.896s    6.124us <- stream_new@src/stream.c:573 task_wakeup
  ssl_sock_io_cb                80581   1.166s    14.47us   200.5ms   2.488us <- sock_conn_iocb@src/sock.c:869 tasklet_wakeup
  h1_io_cb                      80263   1.506s    18.77us   30.08ms   374.0ns <- ssl_sock_io_cb@src/ssl_sock.c:6530 tasklet_wakeup
  h1_io_cb                       8904   4.052ms   455.0ns   346.1ms   38.87us <- h1_takeover@src/mux_h1.c:4124 tasklet_wakeup
  accept_queue_process           2836   22.44ms   7.914us   149.2ms   52.61us <- listener_accept@src/listener.c:1209 tasklet_wakeup
  process_chk                    2776   101.7ms   36.64us   41.24ms   14.86us <- wake_expired_tasks@src/task.c:350 task_wakeup
  h1_io_cb                       2348   8.980ms   3.824us   1.056ms   449.0ns <- conn_subscribe@src/connection.c:736 tasklet_wakeup
  h1_timeout_task                1853   36.71ms   19.81us   27.42ms   14.79us <- wake_expired_tasks@src/task.c:350 task_wakeup
  h1_timeout_task                1265   780.1us   616.0ns   37.71ms   29.81us <- h1_release@src/mux_h1.c:1041 task_wakeup
  ssl_sock_io_cb                  499   856.4ms   1.716ms   14.09ms   28.24us <- ssl_sock_start@src/ssl_sock.c:5864 tasklet_wakeup
  ssl_sock_io_cb                  470   181.0us   385.0ns   3.569ms   7.594us <- conn_subscribe@src/connection.c:736 tasklet_wakeup
  sc_conn_io_cb                   159   1.302ms   8.186us   310.6us   1.953us <- h1_wake_stream_for_send@src/mux_h1.c:2583 tasklet_wakeup
  h1_io_cb                        159   248.4us   1.562us   131.7us   828.0ns <- sock_conn_iocb@src/sock.c:849 tasklet_wakeup
  process_chk                     121   2.392ms   19.77us   675.6us   5.583us <- wake_srv_chk@src/check.c:1055 task_wakeup
  process_chk                      94   73.14us   778.0ns   3.725ms   39.63us <- run_tasks_from_lists@src/task.c:658 task_drop_running
  srv_chk_io_cb                    87   369.2us   4.243us   85.58us   983.0ns <- sock_conn_iocb@src/sock.c:849 tasklet_wakeup
  srv_chk_io_cb                    87   740.0us   8.506us   318.7us   3.662us <- sock_conn_iocb@src/sock.c:869 tasklet_wakeup
  srv_cleanup_toremove_conns       78   1.539ms   19.73us   4.841ms   62.07us <- srv_cleanup_idle_conns@src/server.c:5976 task_wakeup
  srv_cleanup_idle_conns           34   353.6us   10.40us   292.7us   8.608us <- wake_expired_tasks@src/task.c:435 task_drop_running
  task_run_applet                   1   1.641us   1.641us   1.026us   1.026us <- sc_applet_create@src/stconn.c:496 appctx_wakeup
  task_run_applet                   1      -         -      295.0ns   295.0ns <- sc_app_chk_snd_applet@src/stconn.c:1021 appctx_wakeup
  process_chk                       1   14.57us   14.57us   443.0ns   443.0ns <- wake_srv_chk@src/check.c:1048 task_wakeup
  sc_conn_io_cb                     1   4.447us   4.447us   222.0ns   222.0ns <- sock_conn_iocb@src/sock.c:869 tasklet_wakeup
`````

**Comment by ut0mt8** [comment]
seems related to https://github.com/haproxy/haproxy/issues/2046 no ?

**Comment by wtarreau** [comment]
Maybe, I had not noticed you were on 2.7.4, for whatever reason I thought it was latest (might have mixed with another issue). And the commit that was reverted was indeed merged before 2.7.4 so technically you have it.

Yeah it would be worth updating to the latest fixes (2.7.8) to verify if the problem is still there.

**Comment by capflam** [comment]
@ut0mt8 already tested the 2.7.8 but the issue is still there.

**Comment by Darlelet** [comment]
May I suggest another test @ut0mt8?

When you notice an instance going rogue (haproxy @100%), do you think you would be able to temporarily redispatch the load (like redirect or block incoming traffic on this specific instance) without reloading/restarting haproxy to see if haproxy keeps maxing the cpu indefinitely or if it gets back to "normal" by itself as soon as the traffic has dropped?

Because from your task profiling outputs (normal vs sick), they both look very similar and they don't bring that much info unfortunately :/

**Comment by ut0mt8** [comment]
> May I suggest another test @ut0mt8?
> 
> When you notice an instance going rogue (haproxy @100%), do you think you would be able to temporarily redispatch the load (like redirect or block incoming traffic on this specific instance) without reloading/restarting haproxy to see if haproxy keeps maxing the cpu indefinitely or if it gets back to "normal" by itself as soon as the traffic has dropped?
> 
> Because from your task profiling outputs (normal vs sick), they both look very similar and they don't bring that much info unfortunately :/

yup I can. my arch allow that.
let's try it.

**Comment by ut0mt8** [comment]
> @ut0mt8 already tested the 2.7.8 but the issue is still there.

which release do you use on prod? cause unfortunately I have plenty of another battles..

**Comment by wtarreau** [comment]
Let's stay on 2.7.8, no need to reinsert old bugs (and sorry for not finding it the second time, at least I hadn't dreamed, you actually had tested it).

**Comment by ut0mt8** [comment]
> May I suggest another test @ut0mt8?
> 
> When you notice an instance going rogue (haproxy @100%), do you think you would be able to temporarily redispatch the load (like redirect or block incoming traffic on this specific instance) without reloading/restarting haproxy to see if haproxy keeps maxing the cpu indefinitely or if it gets back to "normal" by itself as soon as the traffic has dropped?
> 
> Because from your task profiling outputs (normal vs sick), they both look very similar and they don't bring that much info unfortunately :/

So just tested and haproxy goes down slowly to 0% (it's dns). 
It's surely a rogue client but still very annoying that haproxy cannot handle it.

**Comment by wtarreau** [comment]
BTW I hadn't noticed your yesterday's response yet (sorry, release in progress), but it's quite interesting. The abundance of pool_io in the wakeup cause really makes me think about an uncaught read0. I know it's not easy to enable certain dumps/captures with your traffic, so I'm trying to limit myself. I was thinking about two "show fd" outputs 10 seconds apart, but they'll be huge (40k lines, ~10 MB each). Another test that's not very expensive could be to issue "strace -ttfs200 -o log -p $(pidof haproxy)" for 2 seconds then press Ctrl-C. It will record the syscalls. I'm pretty sure we'll notice multiple calls to recvfrom() or recvmsg() with a result of 0 for the same FD. If so that's definitely a bug! Do not let it run too long, it will slow the process down like crazy, and with your traffic, so 2 seconds should be enough

**Comment by wtarreau** [comment]
Ah so it's not an uncaught read0.

**Comment by wtarreau** [comment]
I've just analysed your "perf top" output. It's puzzling that only libcrypto and malloc are there, neither haproxy nor the kenrel. It could really be the result of clients hammering it. And since you're on OpenSSL 3.0, we know that this version's performance is abysmal, which could definitely be confirmed by the "perf top" output. If you built it yourself (which I suspect based on the version number), you may be interested in trying again with 1.1.1 which is 5-20 times faster depending on the usage pattern.

**Comment by ut0mt8** [comment]
> I've just analysed your "perf top" output. It's puzzling that only libcrypto and malloc are there, neither haproxy nor the kenrel. It could really be the result of clients hammering it. And since you're on OpenSSL 3.0, we know that this version's performance is abysmal, which could definitely be confirmed by the "perf top" output. If you built it yourself (which I suspect based on the version number), you may be interested in trying again with 1.1.1 which is 5-20 times faster depending on the usage pattern.

well I can recompile and redeploy with openssl 1.1 no problem (idk why I changed that ?)

**Comment by wtarreau** [comment]
Probably because you believed it was better. Rest assured you're not the only one, many high-traffic users are getting caught by this one when migrating from older LTS distros to newer ones. There's even a meta-issue on the openssl github to collect as much as possible of these 3.0 massive perf regression reports.

**Comment by ut0mt8** [comment]
OK so I deployed a version with openssl1.11. still have the problem but at least perf output seems more reasonable:

`````
46.82%  libcrypto.so.1.1    [.] __bn_sqr8x_mont
  14.02%  libcrypto.so.1.1    [.] __bn_mul4x_mont
   4.40%  libcrypto.so.1.1    [.] MOD_EXP_CTIME_COPY_FROM_PREBUF
   1.75%  libcrypto.so.1.1    [.] fe51_mul
   1.58%  libc.so.6           [.] _int_malloc
   1.08%  libcrypto.so.1.1    [.] fe51_sq
   0.73%  libcrypto.so.1.1    [.] sha512_block_data_order
   0.61%  libc.so.6           [.] _int_free
   0.59%  libc.so.6           [.] malloc
   0.50%  libcrypto.so.1.1    [.] fe_mul
   0.44%  [kernel]            [k] __fget_light
   0.40%  [kernel]            [k] __softirqentry_text_start
   0.38%  [kernel]            [k] kmem_cache_free
   0.37%  [kernel]            [k] do_epoll_ctl
   0.34%  [kernel]            [k] __inet_lookup_established
   0.33%  [kernel]            [k] sock_poll
   0.33%  libc.so.6           [.] __GI___memcpy_simd
   0.31%  haproxy             [.] h1_headers_to_hdr_list
   0.31%  libcrypto.so.1.1    [.] x25519_scalar_mult
   0.30%  haproxy             [.] process_stream
   0.29%  [kernel]            [k] __wake_up_common_lock
   0.29%  haproxy             [.] sock_conn_iocb
   0.28%  haproxy             [.] sc_conn_sync_recv
   0.27%  [kernel]            [k] __arch_copy_to_user
   0.26%  haproxy             [.] h1_snd_buf
   0.26%  libc.so.6           [.] cfree@GLIBC_2.17
   0.26%  [kernel]            [k] ena_start_xmit
   0.25%  haproxy             [.] pool_evict_last_items.constprop.0
   0.24%  libcrypto.so.1.1    [.] bn_mul_mont_fixed_top
   0.23%  [kernel]            [k] sch_direct_xmit
   0.22%  libc.so.6           [.] unlink_chunk.constprop.0
   0.22%  [kernel]            [k] arch_local_irq_restore
   0.22%  [kernel]            [k] el0_svc_common.constprop.0
   0.21%  [kernel]            [k] memmove
   0.21%  haproxy             [.] run_tasks_from_lists
   0.20%  libcrypto.so.1.1    [.] fe_sq
   0.19%  [kernel]            [k] __mod_timer
   0.19%  [kernel]            [k] ena_clean_tx_irq
   0.19%  [kernel]            [k] arch_local_irq_restore
   0.18%  libcrypto.so.1.1    [.] OPENSSL_cleanse
   0.18%  haproxy             [.] eb32_insert
   0.17%  [kernel]            [k] selinux_socket_sendmsg
   0.17%  libc.so.6           [.] malloc_consolidate
   0.17%  haproxy             [.] __pool_free
   0.16%  [kernel]            [k] skb_release_data
   0.16%  haproxy             [.] pool_get_from_os
   0.15%  libcrypto.so.1.1    [.] BN_mod_exp_mont_consttime
   0.15%  haproxy             [.] h1_refresh_timeout
   0.15%  [kernel]            [k] arch_local_irq_enable
   0.15%  [kernel]            [k] __slab_free
   0.14%  [kernel]            [k] tcp_ack
   0.14%  haproxy             [.] sc_conn_recv
   0.14%  haproxy             [.] __http_find_header
`````

and

``````
[root@ip-10-80-250-243 ec2-user]# echo "show profiling tasks" | socat stdio unix-connect:/run/haproxy-admin.sock
Tasks activity:
  function                      calls   cpu_tot   cpu_avg   lat_tot   lat_avg
  process_stream               234894   3.902s    16.61us   1.738h    26.64ms <- sc_notify@src/stconn.c:1234 task_wakeup
  h1_io_cb                     232890   2.047s    8.788us   28.21s    121.1us <- sock_conn_iocb@src/sock.c:869 tasklet_wakeup
  sc_conn_io_cb                180444   411.4ms   2.280us   1.157h    23.09ms <- sc_app_chk_rcv_conn@src/stconn.c:789 tasklet_wakeup
  h1_timeout_task              117615   39.71ms   337.0ns   53.06m    27.07ms <- h1_release@src/mux_h1.c:1041 task_wakeup
  sc_conn_io_cb                117348   1.042s    8.878us   1.416m    723.8us <- h1_wake_stream_for_send@src/mux_h1.c:2583 tasklet_wakeup
  h1_io_cb                     117337   206.2ms   1.757us   27.84s    237.3us <- sock_conn_iocb@src/sock.c:849 tasklet_wakeup
  process_stream               117054   2.656s    22.69us   49.73m    25.49ms <- stream_new@src/stream.c:573 task_wakeup
  session_expire_embryonic      28889   513.8ms   17.79us   11.89m    24.69ms <- wake_expired_tasks@src/task.c:350 task_wakeup
  other                         12305      -         -      5.182m    25.27ms <- wake_expired_tasks@src/task.c:350 task_wakeup
  accept_queue_process          11754   311.3ms   26.48us   30.06s    2.557ms <- listener_accept@src/listener.c:1209 tasklet_wakeup
  accept_queue_process           5581   120.8ms   21.64us   9.631s    1.726ms <- accept_queue_process@src/listener.c:166 tasklet_wakeup
  h1_io_cb                       3073   56.76ms   18.47us   4.192s    1.364ms <- conn_subscribe@src/connection.c:736 tasklet_wakeup
  h1_timeout_task                1961   28.14ms   14.35us   49.04s    25.01ms <- wake_expired_tasks@src/task.c:350 task_wakeup
  sc_conn_io_cb                  1806   6.954ms   3.850us   11.34s    6.278ms <- h1_wake_stream_for_recv@src/mux_h1.c:2573 tasklet_wakeup
  process_chk                    1031   23.70ms   22.99us   25.69s    24.92ms <- wake_expired_tasks@src/task.c:350 task_wakeup
  ssl_sock_io_cb                  194   193.4us   997.0ns   3.754s    19.35ms <- sock_conn_iocb@src/sock.c:869 tasklet_wakeup
  h1_io_cb                        192   7.225ms   37.63us   4.344s    22.63ms <- ssl_sock_io_cb@src/ssl_sock.c:6530 tasklet_wakeup
  process_chk                      51   652.0us   12.79us   1.343s    26.34ms <- wake_srv_chk@src/check.c:1055 task_wakeup
  srv_chk_io_cb                    36   214.6us   5.959us   6.345ms   176.2us <- sock_conn_iocb@src/sock.c:849 tasklet_wakeup
  srv_chk_io_cb                    36   283.8us   7.882us   2.973ms   82.58us <- sock_conn_iocb@src/sock.c:869 tasklet_wakeup
  process_stream                   19   316.2us   16.64us   423.9ms   22.31ms <- wake_expired_tasks@src/task.c:350 task_wakeup
  task_run_applet                   3   1.453ms   484.4us   515.7us   171.9us <- sc_applet_create@src/stconn.c:496 appctx_wakeup
  task_run_applet                   2   2.429us   1.214us   9.745ms   4.873ms <- run_tasks_from_lists@src/task.c:658 task_drop_running
  sc_conn_io_cb                     1   3.192us   3.192us   5.263ms   5.263ms <- sock_conn_iocb@src/sock.c:869 tasklet_wakeup


``````

**Comment by wtarreau** [comment]
What it shows is that your machine is killed by TLS handshakes. It seems that your users do not even support keep-alive nor TLS resume. Maybe it's specific to a certain type of client you're dealing with. Would your certificates be 4096 bits by any chance ? These ones are extremely CPU-intensive, you'll roughly count on 300-500 conns/s/core depending on the CPU frequency. Do you have a way to send less traffic there to observe how much remains reasonable per node ? Alternately, do you know if your clients support ECDSA certificates ? They're roughly 5 times less CPU intensive.

**Comment by chipitsine** [comment]
I wonder whether openssl is built with or without asm (assembler) support.

```
46.82%  libcrypto.so.1.1    [.] __bn_sqr8x_mont
  14.02%  libcrypto.so.1.1    [.] __bn_mul4x_mont
```

looks too huge even for full tls handshakes

**Comment by wtarreau** [comment]
Interesting. I'm seeing this on my machine with 1.1.1 and full TLS handshakes:
```
  18.56%  haproxy                  [.] __bn_sqrx8x_reduction
  11.38%  haproxy                  [.] bn_sqrx8x_internal
  11.16%  haproxy                  [.] mulx4x_internal
   2.32%  haproxy                  [.] sha512_block_data_order_avx2
   1.49%  haproxy                  [.] fe_mul
   1.41%  haproxy                  [.] __bn_postx4x_internal
   1.17%  libc-2.33.so             [.] malloc
   1.12%  haproxy                  [.] x25519_fe64_sqr
   0.96%  libpthread-2.33.so       [.] __pthread_rwlock_rdlock
   0.96%  libc-2.33.so             [.] _int_malloc
   0.90%  libpthread-2.33.so       [.] __pthread_rwlock_unlock
   0.88%  libc-2.33.so             [.] _int_free
   0.82%  libc-2.33.so             [.] __strcmp_avx2
   0.71%  libc-2.33.so             [.] cfree@GLIBC_2.2.5
```

__bn_sqr8x_mont seems to only be reported on arm64.

Oh, wait a minute, I hadn't noticed, but it actually *is* aarch64! That makes a huge difference, aarch64 is way less powerful than x86 at TLS handshakes. On my machine (LX2), 16 cores at 2.2 GHz only deliver 2950 handshakes per second in 2048 bits, that's 184 handshakes per second per core or 83 handshakes per second per core per GHz!
That's about the perf I achieve on x86 with only 2 cores at 3.5!

This if you have to deal with many TLS handshakes due to clients not supporting resume, just give up on ARM, I'm sorry. These platforms are amazing for a lot of things and for their efficiency, but TLS handhsakes do not scale. I have no experience with ECDSA there though, it could be worth trying.

**Comment by ut0mt8** [comment]
> Interesting. I'm seeing this on my machine with 1.1.1 and full TLS handshakes:
> 
> ```
>   18.56%  haproxy                  [.] __bn_sqrx8x_reduction
>   11.38%  haproxy                  [.] bn_sqrx8x_internal
>   11.16%  haproxy                  [.] mulx4x_internal
>    2.32%  haproxy                  [.] sha512_block_data_order_avx2
>    1.49%  haproxy                  [.] fe_mul
>    1.41%  haproxy                  [.] __bn_postx4x_internal
>    1.17%  libc-2.33.so             [.] malloc
>    1.12%  haproxy                  [.] x25519_fe64_sqr
>    0.96%  libpthread-2.33.so       [.] __pthread_rwlock_rdlock
>    0.96%  libc-2.33.so             [.] _int_malloc
>    0.90%  libpthread-2.33.so       [.] __pthread_rwlock_unlock
>    0.88%  libc-2.33.so             [.] _int_free
>    0.82%  libc-2.33.so             [.] __strcmp_avx2
>    0.71%  libc-2.33.so             [.] cfree@GLIBC_2.2.5
> ```
> 
> __bn_sqr8x_mont seems to only be reported on arm64.
> 
> Oh, wait a minute, I hadn't noticed, but it actually _is_ aarch64! That makes a huge difference, aarch64 is way less powerful than x86 at TLS handshakes. On my machine (LX2), 16 cores at 2.2 GHz only deliver 2950 handshakes per second in 2048 bits, that's 184 handshakes per second per core or 83 handshakes per second per core per GHz! That's about the perf I achieve on x86 with only 2 cores at 3.5!
> 
> This if you have to deal with many TLS handshakes due to clients not supporting resume, just give up on ARM, I'm sorry. These platforms are amazing for a lot of things and for their efficiency, but TLS handhsakes do not scale. I have no experience with ECDSA there though, it could be worth trying.

Just read that from openssl mailing list :/ my big plan to put everything on graviton instance bite the dust....
let's switch back to x64.

**Comment by chipitsine** [comment]
I recall aarch64 (in my perf tests) were about 20% ahead of xeon (per rack unit, taking into account that TLS is scaled over all cores and aarch64 have really lot of cores)

**Comment by chipitsine** [comment]
also, I found significant gap in aarch64 gcc compilers, if you use gcc13 it can give 10% perf compared with gcc9

**Comment by ut0mt8** [comment]
> also, I found significant gap in aarch64 gcc compilers, if you use gcc13 it can give 10% perf compared with gcc9

gcc11.3.1 here

**Comment by wtarreau** [comment]
In fact on x86 the MUL instruction was specifically improved to chain such operations to perform large multiplies extremely fast. And one specific good feature of x86 is that the mul is 64x64->128. On the opposite, on ARM (which is a RISC machine) it's doing 64x64->64 so you have to perform way more operations there for the same result since in practice you're constrained to 32x32->64 if you don't want to overflow.

**Comment by wtarreau** [comment]
@ut0mt8 BTW is it a graviton or graviton2 ? The graviton2 is a Neoverse-N1 and is amazingly fast. I have not run TLS handshakes benchmarks on it though, but given your numbers are still above mine (I'm on A72, same as Graviton) I would suspect you're already on that one.

**Comment by chipitsine** [comment]
@ut0mt8 , is that server publicly available ?

I would have a look at https://www.ssllabs.com/ssltest/analyze.html?d=star.dsp.corp.com

(if it is not available, can you test it using https://testssl.sh/ ?)

**Comment by ut0mt8** [comment]
> @ut0mt8 BTW is it a graviton or graviton2 ? The graviton2 is a Neoverse-N1 and is amazingly fast. I have not run TLS handshakes benchmarks on it though, but given your numbers are still above mine (I'm on A72, same as Graviton) I would suspect you're already on that one.

this one is on graviton "1". graviton 2 wasn't available at time but let's change that.

**Comment by ut0mt8** [comment]
> @ut0mt8 , is that server publicly available ?
> 
> I would have a look at https://www.ssllabs.com/ssltest/analyze.html?d=star.dsp.corp.com
> 
> (if it is not available, can you test it using https://testssl.sh/ ?)

https://www.ssllabs.com/ssltest/analyze.html?d=adx-ap.dsp.adikteev.com ; I don't see anything obviously wrong (btw security is really not the priority here, and if my peers can contact me in plain http I would be happy; unfortunately some SSPs won't)

**Comment by chipitsine** [comment]
Once I profiled cpu usage on "internet ads" application, clients connect only once and every ssl handshake is full. From my observation the major CPU eater was ECDHE exchange (required for forward secrecy).

so ... bad advice from my (if your customers are ok with plain http as you said)

disable TLS1.3 and leave only light weight GCM suites without ECDHE

![image](https://github.com/haproxy/haproxy/assets/2217296/90da8862-b930-4ab1-bf13-bf649086cbbf)

**Comment by chipitsine** [comment]
also, "OCSP stapling = No", that means client may perform additional round trip for OCSP.
I'd enable it

**Comment by chipitsine** [comment]
another interesting thing is "tls resumption"

![image](https://github.com/haproxy/haproxy/assets/2217296/c7b77187-78bb-46ce-a608-1d1e7e4acd61)


you say that you observe cpu spikes after reload, it looks like ssl storm indeed. but it should not happen if there's tickets persistance (not sure actually whether your clients support it or not).

do you have tickets persistance https://docs.haproxy.org/2.7/configuration.html#tls-ticket-keys ?

(if configured, tickets should survive across reload)

**Comment by lukastribus** [comment]
> observation the major CPU eater was ECDHE exchange (required for forward secrecy).

Are you sure you are not confusing ECDHE with DHE?

DHE is very heavy on server CPUs and it's really not necessary for FS anymore, since everyone properly implements ECDHE.

Disabling DHE, but not ECDHE, helps. Switching from RSA to EC certificates helps, at least provide both certificate types, so most clients use the EC cert, lowering CPU load.

A few dated, but specific benchmarks about this are here:

https://vincent.bernat.ch/en/blog/2011-ssl-dos-mitigation
https://vincent.bernat.ch/en/blog/2011-ssl-perfect-forward-secrecy


TLS1.3 and ECDHE is very usable at scale.



> you say that you observe cpu spikes after reload, it looks like ssl storm indeed. but it should not happen if there's tickets persistance (not sure actually whether your clients support it or not).
> 
> do you have tickets persistance https://docs.haproxy.org/2.7/configuration.html#tls-ticket-keys ?
> 
> (if configured, tickets should survive across reload)

Good point, either properly configure tls-tickets with files or try to disable tickets with `no-tls-tickets` ; otherwise when reloading it can indeed cause a negotiation storm due to clients unable to resume the TLS ticket because the runtime tls-ticket key is gone.

**Comment by chipitsine** [comment]
I refer ` ECDH x25519` part. It was the major cpu eater (in my very specific "single shot full handshake" workload). That part does not depend on ECDSA/RSA unfortunately

**Comment by wtarreau** [comment]
By the way regarding the possible TLS storm on reload, that's a good point. If the clients only connect and stay connected as we could expect them to, then `close-spread-time` would definitely help, by progressively closing the old connections and smoothly transferring the load from the old to the new process. From your "show info" we saw something like 40k conns. At say 1k/s you'd need no less than 40 seconds to transfer the load, so setting `close-spread-time 40s` could help. And yes, please try to create ECDSA certificates. They will be faster. By how much I don't know on ARM, but they will definitely be faster.

**Comment by wtarreau** [comment]
Also a 3072 bits cert is certainly expensive on such a machine, one more reason for trying ECDSA.

**Comment by chipitsine** [comment]
I think it would be just wonderful to collect the above mentioned profiling information for different set of settings. It worth keeping them in some public place

**Comment by wtarreau** [comment]
That's among the stuff I'd like to see in the wiki. I.e. share measurements of various common and reproducible metrics on different types of hardware. For example if we see that an A72 delivers 83 cps/GHz/core in RSA2K, that can help people choose their hardware for specific tasks. But it's often extremely difficult to figure what you **need** first. For example the SSL storm during a reload might be the type of situation it's difficult to recover from, because clients that are not served just give up, but their pending connection and handshake are still in the queue and will lead to a calculation before detecting the connection's dead. So there's a difference between sizing for routine usage and sizing for recovery in case of an incident or a manipulation error.

**Comment by ut0mt8** [comment]
I can disable TLS1.3 yes good idea and leave only cheapest cipher. 
Just it's sound like a ssl storm but not after restart or reload.  I don't reload or restart by myself.
It's seems just random. but it can be one of my client which suddenly broke some established connections provoking  this cnx storm. 

this is also what I observe when in struggle : connection rate around 5K/s which seems to much for it...
in normal scenario on this region one instance handle around 400 cnx/s for 38k qps ; and very small load but I have to keep rooms for handling reco. 
In short I have very clients ; and when all are connected all goods; but traffic vary, and instances also (not mentioning I use spot and autoscaling).

let see tomorrow with first c7g instances; and then if not conclusive intel ones.

and maybe I have to split completely SSL trafic from other. (at least two frontend section...)

**Comment by chipitsine** [comment]
> That's among the stuff I'd like to see in the wiki. I.e. share measurements of various common and reproducible metrics on different types of hardware. For example if we see that an A72 delivers 83 cps/GHz/core in RSA2K, that can help people choose their hardware for specific tasks. But it's often extremely difficult to figure what you **need** first. For example the SSL storm during a reload might be the type of situation it's difficult to recover from, because clients that are not served just give up, but their pending connection and handshake are still in the queue and will lead to a calculation before detecting the connection's dead. So there's a difference between sizing for routine usage and sizing for recovery in case of an incident or a manipulation error.

surprisingly enough, but we also found that single xeon core handle 700-1000 full handshakes per sec depending on suite.
maybe collect on wiki something like

1. cpu
2. compiler
3. openssl version
4. suite
5. number of full handshakes per sec per

**Comment by chipitsine** [comment]
> I can disable TLS1.3 yes good idea and leave only cheapest cipher. Just it's sound like a ssl storm but not after restart or reload. I don't reload or restart by myself. It's seems just random. but it can be one of my client which suddenly broke some established connections provoking this cnx storm.
> 
> this is also what I observe when in struggle : connection rate around 5K/s which seems to much for it... in normal scenario on this region one instance handle around 400 cnx/s for 38k qps ; and very small load but I have to keep rooms for handling reco. In short I have very clients ; and when all are connected all goods; but traffic vary, and instances also (not mentioning I use spot and autoscaling).
> 
> let see tomorrow with first c7g instances; and then if not conclusive intel ones.
> 
> and maybe I have to split completely SSL trafic from other. (at least two frontend section...)

sorry, I thought that ssl storm is initiated by haproxy reload.

also, I recall that some unexpected response may lead to unhandled exceptions on client side, thus entire client was restarted, no chance to keep ssl ticket persistance on client side.

not sure how to protect from that :)

**Comment by wtarreau** [comment]
Well, the clients are allowed to restart their own programs as well. That's why it's annoying to be CPU-bound like this. If it causes trouble to all other ones, it's possible to dedicate a frontend and put a `rate-limit sessions` on it to prevent them from going too fast, but it will have an impact on the time it takes for them to recover, even though it will protect other ones.

**Comment by wtarreau** [comment]
A few perf measurements here using openssl speed (sign/s per core):
|   Machine      |   RSA2048   |   RSA3072    |    ECDSAP256 |
|--|--|--|--|
|  Xeon Gold 6212U @2.4 GHz |  1188 | 393  | 32137 |
| i5-6300U @3.0 GHz| 1398 | 466 | 34137 |
| Cortex A72 @2.2 GHz | 245 | 80 | 13610 |
| Cortex A76 @2.3 GHz | 266 | 85 | 17191 |

This can be checked using: `openssl speed rsa2048 rsa3072 ecdsap256` and looking at the `sign/s` column. Of course this doesn't include all the rest, signatures, PFS etc but it gives a good indication of the effect of choosing certain key algorithms and sizes for the certificates (especially since you said that in your case you don't care about the security on this). As you can see RSA3072 is 3 times more expensive than RSA2048, and 100-200 times more expensive than the equivalent ECDSAP256. Thus you should definitely regenerate your certs in 2048 bits and add ECDSA for clients that support it. In this case your Graviton1 might be quite sufficient as you can see (A72 above).

**Comment by wtarreau** [comment]
Ah I had misread, it's already 2048 on your site, the 3072 that was mentioned was for the RSA-equivalent of the ECDHE 256. So the perf ratio between RSA2K and ECDSA256 should still be around 50 on your machine.

**Comment by wtarreau** [comment]
Does this issue still need to stay open ?

**Comment by ut0mt8** [comment]
> Does this issue still need to stay open ?

Well it's not resolved but it's not easy to say if it's really haproxy specific

**Comment by wtarreau** [comment]
OK but did you have a chance to try on a more punchy machine, or using ECDSA certs instead, given that, as shown above, the cortex A72 in Graviton1 is severely limited for RSA handshakes ?

**Comment by wtarreau** [comment]
BTW I can complete the table above with Neoverse-N1 (as found in graviton2), it's 270 at 2.3 GHz or sensibly the same as A76.

**Comment by ut0mt8** [comment]
> OK but did you have a chance to try on a more punchy machine, or using ECDSA certs instead, given that, as shown above, the cortex A72 in Graviton1 is severely limited for RSA handshakes ?

 I tried with Graviton2 and it happen less often but still. And unfortunately my certificate authority seems not handling ecdsa cert (Gandi ?, or at least the process is broken)

**Comment by wtarreau** [comment]
OK so as we previously guessed you're exactly in the situation where an ARM instance is not affordable for your use case, because using x86 for twice the price you'll get 5 times the performance. I think that Gandi supports ECDSA though but as you say there may be something broken somewhere in the process. It could be worth contacting them because if you have to switch to a more expensive x86 just for a matter of poorly documented process, that's a shame!

**Comment by ut0mt8** [comment]
The x86 is unfortunately not that much expensive.

**Comment by wtarreau** [comment]
I guess you meant "fortunately" above :-)

**Comment by ut0mt8** [comment]
Fortunately for us but not for the science ;)

**Comment by wtarreau** [comment]
Well, the science was already covered by the tests above and showed physical limits. Sadly, RSA will always be slow on ARM (and likely most RISC CPUs) because of the way the MUL instruction works. It's made of only a half-multiplier which does 64x64->64 so in practice it's at best usable for 32x32->64 due to the carry, so RSA calculations can only be done 32 bits at a time where on x86 you have 64x64->128 allowing to use the full 64 bits on each pass. In addition, recently x86 multipliers were improved for RSA by supporting pipelined chained multiplications that further speed the operation up. Until ARM decides to create a new instruction involving a 4th register it will continue to lag behind on this type of workload :-(

**Comment by wtarreau** [comment]
Hello,
Should we close this one or are you still facing issues on the x86 instance ?

**Comment by ut0mt8** [comment]
Hi @wtarreau ; unfortunately yes; same symptom; same perf output/ but it's now very rare. Once every two weeks.

**Comment by wtarreau** [comment]
You're probably getting connection peaks from time to time :-/
Are you still on openssl-1.1.1 now or did you switch back to 3.0 ?

**Comment by DemiMarie** [comment]
@wtarreau does this problem also impact elliptic curve cryptography and Poly1305?

**Comment by wtarreau** [comment]
See above, you'll notice that ECDSAP256 is only about half as slow on ARM compared to x86, and numbers are much higher than RSA so the limitations will be elsewhere. No measures for poly1305 though.

**Comment by DemiMarie** [comment]
Thanks @wtarreau!

**Comment by capflam** [comment]
Sorry to ask. Is there something more to do on this issue ? Is it an openssl 3.0 issue or not ?

**Comment by ut0mt8** [comment]
Yes sorry let's close it

**Comment by capflam** [comment]
Ok, no problem,  thanks !
